silhouette/app/com/mohiva/play/silhouette/impl/providers/CredentialsProvider.scala

```diff
@@ -66,16 +64,16 @@ class CredentialsProvider @Inject() (
   def authenticate(credentials: Credentials): Future[LoginInfo] = {
     loginInfo(credentials).flatMap { loginInfo =>
       authInfoRepository.find[PasswordInfo](loginInfo).flatMap {
-        case Some(authInfo) => passwordHasherList.find(_.id == authInfo.hasher) match {
+        case Some(authInfo) => passwordHasherRegistry.find(authInfo) match {
           case Some(hasher) if hasher.matches(authInfo, credentials.password) =>
-            if (hasher != passwordHasher) {
-              authInfoRepository.update(loginInfo, passwordHasher.hash(credentials.password)).map(_ => loginInfo)
+            if (passwordHasherRegistry.isDeprecated(hasher) || hasher.isDeprecated(authInfo).contains(true)) {
+              authInfoRepository.update(loginInfo, passwordHasherRegistry.current.hash(credentials.password)).map(_ => loginInfo)
             } else {
               Future.successful(loginInfo)
             }
           case Some(hasher) => throw new InvalidPasswordException(InvalidPassword.format(id))
           case None => throw new ConfigurationException(UnsupportedHasher.format(
-            id, authInfo.hasher, passwordHasherList.map(_.id).mkString(", ")
+            id, authInfo.hasher, passwordHasherRegistry.all.map(_.id).mkString(", ")
```

Comment:
I see you had to make the exact same changes twice. Keep in mind my suggestion to refactor the greater bulk of these two classes into one.